### ISSUE TITLE
Crash in CollisionPolygon2D::_update_xform_in_parent() in 2.1 branch (regression)

### ISSUE THREAD

**Issue by akien-mga** [godotengine/godot#17780]
**Godot version:**
Godot 2.1.5 RC1 / 2.1 branch (3764cf6)

**OS/device including version:**
Any.

**Issue description:**
@avencherus sent me a project which crashes on load in Godot 2.1.5 RC1 while it runs fine in 2.1.4-stable. It's not a public project so I'll see if I can find the fix myself, or isolate the issue and make a reproducer.

CC @RandomShaper as I suspect #15607 caused the issue.

Stacktrace:
```
[1] /lib64/libc.so.6(+0x338a0) [0x7f5b81ab08a0] (??:0)
[2] Vector<CollisionObject2D::ShapeData>::_get_size() const (/home/akien/Projects/godot/godot-2.1/core/vector.h:60)
[3] Vector<CollisionObject2D::ShapeData>::size() const (/home/akien/Projects/godot/godot-2.1/core/vector.h:111)
[4] CollisionObject2D::set_shape_transform(int, Matrix32 const&) (/home/akien/Projects/godot/godot-2.1/scene/2d/collision_object_2d.cpp:279 (discriminator 2))
[5] CollisionPolygon2D::_update_xform_in_parent() (/home/akien/Projects/godot/godot-2.1/scene/2d/collision_polygon_2d.cpp:101)
[6] CollisionPolygon2D::_notification(int) (/home/akien/Projects/godot/godot-2.1/scene/2d/collision_polygon_2d.cpp:174)
[7] CollisionPolygon2D::_notificationv(int, bool) (/home/akien/Projects/godot/godot-2.1/scene/2d/collision_polygon_2d.h:38 (discriminator 14))
[8] Object::notification(int, bool) (/home/akien/Projects/godot/godot-2.1/core/object.cpp:865)
[9] Node::_propagate_enter_tree() (/home/akien/Projects/godot/godot-2.1/scene/main/node.cpp:173)
[10] Node::_propagate_enter_tree() (/home/akien/Projects/godot/godot-2.1/scene/main/node.cpp:184 (discriminator 2))
[11] Node::_propagate_enter_tree() (/home/akien/Projects/godot/godot-2.1/scene/main/node.cpp:184 (discriminator 2))
[12] Node::_propagate_enter_tree() (/home/akien/Projects/godot/godot-2.1/scene/main/node.cpp:184 (discriminator 2))
[13] Node::_set_tree(SceneTree*) (/home/akien/Projects/godot/godot-2.1/scene/main/node.cpp:1928)
[14] Node::_add_child_nocheck(Node*, StringName const&) (/home/akien/Projects/godot/godot-2.1/scene/main/node.cpp:698)
[15] Node::add_child(Node*, bool) (/home/akien/Projects/godot/godot-2.1/scene/main/node.cpp:728 (discriminator 2))
[16] MethodBind2<Node*, bool>::call(Object*, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot-2.1/core/method_bind.gen.inc:1275 (discriminator 19))
[17] Object::call(StringName const&, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot-2.1/core/object.cpp:853 (discriminator 1))
[18] Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) (/home/akien/Projects/godot/godot-2.1/core/variant_call.cpp:991 (discriminator 1))
[19] GDFunction::call(GDInstance*, Variant const**, int, Variant::CallError&, GDFunction::CallState*) (/home/akien/Projects/godot/godot-2.1/modules/gdscript/gd_function.cpp:641)
[20] GDInstance::call(StringName const&, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot-2.1/modules/gdscript/gd_script.cpp:1112)
[21] Object::call(StringName const&, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot-2.1/core/object.cpp:832 (discriminator 1))
[22] Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) (/home/akien/Projects/godot/godot-2.1/core/variant_call.cpp:991 (discriminator 1))
[23] GDFunction::call(GDInstance*, Variant const**, int, Variant::CallError&, GDFunction::CallState*) (/home/akien/Projects/godot/godot-2.1/modules/gdscript/gd_function.cpp:641)
[24] GDInstance::call(StringName const&, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot-2.1/modules/gdscript/gd_script.cpp:1112)
[25] Object::call(StringName const&, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot-2.1/core/object.cpp:832 (discriminator 1))
[26] Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) (/home/akien/Projects/godot/godot-2.1/core/variant_call.cpp:991 (discriminator 1))
[27] GDFunction::call(GDInstance*, Variant const**, int, Variant::CallError&, GDFunction::CallState*) (/home/akien/Projects/godot/godot-2.1/modules/gdscript/gd_function.cpp:641)
[28] GDFunction::call(GDInstance*, Variant const**, int, Variant::CallError&, GDFunction::CallState*) (/home/akien/Projects/godot/godot-2.1/modules/gdscript/gd_function.cpp:762)
[29] GDInstance::call(StringName const&, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot-2.1/modules/gdscript/gd_script.cpp:1112)
[30] Object::call(StringName const&, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot-2.1/core/object.cpp:832 (discriminator 1))
[31] Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) (/home/akien/Projects/godot/godot-2.1/core/variant_call.cpp:991 (discriminator 1))
[32] GDFunction::call(GDInstance*, Variant const**, int, Variant::CallError&, GDFunction::CallState*) (/home/akien/Projects/godot/godot-2.1/modules/gdscript/gd_function.cpp:641)
[33] GDFunctionState::_signal_callback(Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot-2.1/modules/gdscript/gd_function.cpp:1359)
[34] MethodBindNative<GDFunctionState>::call(Object*, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot-2.1/core/method_bind.h:282 (discriminator 4))
[35] Object::call(StringName const&, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot-2.1/core/object.cpp:853 (discriminator 1))
[36] Object::emit_signal(StringName const&, Variant const**, int) (/home/akien/Projects/godot/godot-2.1/core/object.cpp:1156 (discriminator 1))
[37] Object::emit_signal(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) (/home/akien/Projects/godot/godot-2.1/core/object.cpp:1197)
[38] SceneTree::idle(float) (/home/akien/Projects/godot/godot-2.1/scene/main/scene_main_loop.cpp:508 (discriminator 2))
[39] Main::iteration() (/home/akien/Projects/godot/godot-2.1/main/main.cpp:1549)
[40] OS_X11::run() (/home/akien/Projects/godot/godot-2.1/platform/x11/os_x11.cpp:2284)
[41] godot-2.1(main+0xa6) [0xc2df1b] (/home/akien/Projects/godot/godot-2.1/platform/x11/godot_x11.cpp:52)
[42] /lib64/libc.so.6(__libc_start_main+0xf0) [0x7f5b81a9d600] (??:0)
[43] godot-2.1(_start+0x29) [0xc2dd89] (/home/iurt/rpmbuild/BUILD/glibc-2.22/csu/../sysdeps/x86_64/start.S:121)
-- END OF BACKTRACE --
Abandon (core dumped)
```

**Comment by akien-mga** [comment]
From a quick glance it seems that deleting those "orphan" collision shapes in the player scenes prevents the crash (but obviously breaks the logic as those are used at runtime):
![screenshot_20180326_180233](https://user-images.githubusercontent.com/4701338/37917869-dc1dc3d0-311f-11e8-9da0-f5dd932b40a4.png)

Since those CollisionPolygon2D's have no parent, I guess the method crashes here since no validation is done on the parent:
https://github.com/godotengine/godot/pull/15607/commits/d7195c18d7e430228db0c916f2af4f0a4ffeda64#diff-90d27e5985a7771fb7e81ed5d76ae317R102

**Comment by avencherus** [comment]
@akien-mga Thanks a ton for tracking that down, it is quite a relief.  It definitely would be something above my current ability to locate.

The characters dynamically change their hit areas to match their visuals, and you're right, I've been using them as references in code, for great time saving effect.  I've been able to fine tune and easily update them visually in the editor, and in code just copy and store the shapes and transforms at initialization.

It's not a problem to modify the approach, the current one was on the spot and certainly hacky, and I should probably adhere to a more technically intended method and tie together multiple Area2Ds instead.

**Comment by avencherus** [comment]
@akien-mga I should've added that I also verified it on my end in 2.1.5 RC1, by adding them to unused Area2D and fixing up the path references.   As far as I can tell the crashing has been resolved from that.

**Comment by akien-mga** [comment]
Yeah your use case is a bit special, but it shouldn't crash nevertheless. If you manage to make a simple project that reproduces this crash, that would help making this code more robust to fix it.

**Comment by avencherus** [comment]
I see what you mean, it isn't so straight forward to reproduce.  I tried to recreate the core steps of that RigidBody2D character scene, but haven't had any luck yet.  If I somehow manage to stumble upon the magic combination and can isolate it, I'll post back.  Though I'm out of ideas at the moment.

**Comment by RandomShaper** [comment]
Please give #17804 a try.

**Comment by akien-mga** [comment]
Fixed by #17804.
